Introduce AnswerStoreUpdater

MebinAbraham opened the pull request:
### What is the context of this PR?
Introduced a new component to handle logic around what happens when the answer store gets updated. I'm hoping this:

- Makes the code more understandable
- Better defines the responsibilities of this object
- Provides a component that can be better tested as a single unit
- Allows a few of the schema-specific integration tests to be removed

Since David's initial additions, I have also moved action relating to the household/relationship from the `questionnaire` view into the `AnswerStoreUpdater` also. I've kept the original commits so you can compare.

### How to review 
- All tests should pass
- Code coverage should not reduce
- No changes to functionality

MebinAbraham commented on tests/app/questionnaire/test_answer_store_updater.py:
```diff
@@ -0,0 +1,249 @@
+# import unittest
```
I was debating whether I should keep this as well?. When modifying these to work with the current version of master, I seemed to end up mocking a lot on top of what's already mocked and seemed almost pointless.

The scenarios are covered via integration tests for the questionnaire.py

nealedj replied:
My 2p...yeah I know what you mean. If high-level integration tests are present it's tempting to just add lots of them because they all end up exercising their dependencies anyway.

I guess a few arguments in favour of keeping these unit tests:

- they will run more quickly than integration tests
- they continue a useful convention where most python modules in survey runner have a corresponding set of unit tests
- it should be quicker/easier to work on the answer store updater module with a dedicated set of tests to verify changes. I.e. new scenarios

ajmaddaford commented on app/questionnaire/answer_store_updater.py:
```diff
@@ -0,0 +1,163 @@
+        else:
+            self._update_questionnaire_store_with_form_data(form.data)
+
+        self._questionnaire_store.add_or_update()
```
We could use the type of form here to determine which method to call here, something like:
```python
if isinstance(form, QuestionnaireForm):
    self._update_questionnaire_store_with_form_data(form.data)
else:
    self._update_questionnaire_store_with_answer_data(form.serialise())
```

ajmaddaford commented on app/questionnaire/answer_store_updater.py:
```diff
@@ -0,0 +1,163 @@
+            self._answer_store.add_or_update(answer)
+
+        if self._current_location not in self._questionnaire_store.completed_blocks:
+            self._questionnaire_store.completed_blocks.append(self._current_location)
```
These two lines are duplicated in both `_update_questionnaire_store_with_` methods and could be moved to the `save_form` method.

ajmaddaford commented on app/questionnaire/answer_store_updater.py:
```diff
@@ -0,0 +1,163 @@
+        self._questionnaire_store = questionnaire_store
+        self._answer_store = self._questionnaire_store.answer_store
+
+    def save_form(self, form):
```
I think this method would be clearer if it was called `save_answers`

dwyeradam commented on app/views/questionnaire.py:
```diff
@@ -200,15 +188,17 @@ def post_block(routing_path, schema, metadata, collection_metadata, answer_store
 
-    disable_mandatory = any(x in request.form for x in ['action[add_answer]', 'action[remove_answer]', 'action[save_sign_out]'])
+    if household_answer_changed:
+        answer_store_updater.remove_repeating_on_household_answers()
```
Can the logic around household answers changing (3 lines) all be dealt with with answer_store_updater? At the moment all the work is done there but the logic to call it sits outside.

MebinAbraham commented:
@ajmaddaford @dwyeradam Ready for re-review.

ajmaddaford commented on tests/app/questionnaire/test_answer_store_updater.py:
```diff
@@ -0,0 +1,266 @@
+        answer_calls = map(mock.call, answers)
+        self.answer_store.add_or_update.assert_has_calls(answer_calls, any_order=True)
+
+    def test_save_form_with_form_data(self):
```
All `test_save_form_*` methods should be renamed `test_save_answers_*` to match the code change.

ajmaddaford commented on tests/app/questionnaire/test_answer_store_updater.py:
```diff
@@ -0,0 +1,266 @@
+
+        form = MagicMock(spec=QuestionnaireForm, data={answer_id: answer_value})
+
+        # @patch('')
```
This should be removed.

ajmaddaford commented on tests/integration/views/test_questionnaire.py:
```diff
@@ -34,17 +36,16 @@ def tearDown(self):
         with self._application.test_request_context():
-            update_questionnaire_store_with_form_data(self.question_store, location, form_data, schema)
+            answer_store_updater = AnswerStoreUpdater(location, schema, self.question_store)
+            answer_store_updater.save_answers(form_data)
```
Is this test relevant now that we have `AnswerStoreUpdater` and it's tests. This looks like it is duplicating tests.

ajmaddaford commented on tests/integration/views/test_questionnaire.py:
```diff
@@ -57,18 +58,17 @@ def test_update_questionnaire_store_with_form_data(self):
         with self._application.test_request_context():
-            update_questionnaire_store_with_form_data(self.question_store, location, form_data, schema)
+            answer_store_updater = AnswerStoreUpdater(location, schema, self.question_store)
+            answer_store_updater.save_answers(form_data)
```
Same comment as above about duplication of tests.

ajmaddaford commented on tests/integration/views/test_questionnaire.py:
```diff
@@ -81,54 +81,64 @@ def test_update_questionnaire_store_with_default_value(self):
         with self._application.test_request_context():
-            update_questionnaire_store_with_answer_data(self.question_store, location, answers, schema)
+            answer_store_updater = AnswerStoreUpdater(location, schema, self.question_store)
+            answer_store_updater.save_answers(form)
```
Same comment as above about duplication of tests. Looking at the rest of these tests the same applies. If the functionality is performed by `AnswerStoreUpdater`, that's where the tests are/should be.

MebinAbraham commented:
Just converting some of the existing integration tests relating to `AnswerStoreUpdater` from `questionnaire.py` to unit tests.

MebinAbraham commented:
@ajmaddaford @dwyeradam Have removed tests from `questionnaire.py` which now fits into `AnswerStoreUpdater`.

bitdivision commented on tests/app/questionnaire/test_answer_store_updater.py:
```diff
@@ -0,0 +1,506 @@
+
+        self.answer_store_updater.save_answers(form)
+
+        self.answer_store.filter.return_value = iter(i[0][0].__dict__ for i in self.answer_store.add_or_update.call_args_list)
```
🙀

MebinAbraham replied:
👀

ajmaddaford replied:
We shouldn't need to do it this way, we  just need a mock for the filter method to return the answers i.e. we can remove the `save_answers` and `add_or_update stuff`.

MebinAbraham replied:
Have updated the unit test to match this.

ajmaddaford commented on tests/app/questionnaire/test_answer_store_updater.py:
```diff
@@ -161,54 +161,53 @@ def test_save_answers_data_with_default_value(self):
-                answer_instance=1,
-                value=''
-            )
+        empty_household_answer = [
```
minor, this should be _answer**s**

Commit:
Introduced AnswerStoreUpdater

- A component responsible for any actions that need to happen as a result of updating the answer store
